IdentityCore/src/cache/mac/MSIDMacCredentialStorageItem.m

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+
+#import "MSIDMacCredentialStorageItem.h"
+
+static NSString *keyDelimiter = @"-";
+
+@interface MSIDMacCredentialStorageItem ()
+
+@property (nonatomic) NSMutableDictionary *cacheObjects;
+@property (nonatomic) dispatch_queue_t queue;
+
+@end
+
+@implementation MSIDMacCredentialStorageItem
+
+- (instancetype)init
+{
+    if(self = [super init])
+    {
+        self.cacheObjects = [NSMutableDictionary dictionary];
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.universalstorage-%@", [NSUUID UUID].UUIDString];
+        self.queue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
+    }
+    
+    return self;
+}
+
+- (void)storeCredential:(MSIDCredentialCacheItem *)credential forKey:(NSString *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects setObject:credential forKey:key];
+    });
+}
+
+- (void)mergeStorageItem:(MSIDMacCredentialStorageItem *)storageItem
+{
+    NSArray *keys = [storageItem allKeys];
+    for (NSString *key in keys)
+    {
+        if (![self storedCredentialForKey:key])
+        {
+            MSIDCredentialCacheItem *credential = [storageItem.cacheObjects objectForKey:key];
+            if (credential)
+            {
+                [self storeCredential:credential forKey:key];
+            }
+        }
+    }
+}
+
+- (MSIDCredentialCacheItem *)storedCredentialForKey:(NSString *)key
+{
+    __block MSIDCredentialCacheItem *credential = nil;
+    
+    dispatch_sync(self.queue, ^{
+        credential = [self.cacheObjects objectForKey:key];
+    });
+    
+    return credential;
+}
+
+- (void)removeStoredCredentialForKey:(NSString *)key
```

Comment:
seems it is safer to do a nil check for `key` :)

Reply:
Key is now created from defaultCacheKey and would never be nil.